File: docs/python-reference/nano_kava/chat_digest.py
"""
file: docs/python-reference/nano_kava/chat_digest.py
author: Stephen Boyett

description:
    Python replica of src/components/chat/transport/chatDigest.js — the browser half that packs a
    Sol conversation for the team digest. `navigator.sendBeacon` has no Python equivalent and is
    omitted; what is here is the packing and the floor that stops a silent visit from mailing
    anyone.

See Also:
    src/components/chat/transport/chatDigest.js
    docs/python-reference/nano_kava/digest.py

---
Copyright © 2026 Cannasol Technologies LLC. All Rights Reserved.
---
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .jsisms import falsy_or, is_truthy

MIN_VISITOR_TURNS = 1
IDLE_TIMEOUT_MS = 3 * 60 * 1000

CONTACT_KEYS = ("name", "company", "email", "phone")


@dataclass
class PanelMessage:
    """JS: the in-panel message object. A `lead` carries `fields` instead of `text`."""

    role: str
    text: str = ""
    fields: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class WireMessage:
    role: str
    text: str


@dataclass(frozen=True)
class OutboundDigest:
    messages: tuple[WireMessage, ...]
    contact: dict[str, str]
    page: str
    lead_sent: bool
    share_authorized: bool
    reason: str


def extract_contact(messages: Any) -> dict[str, str]:
    """
    The lead card holds the best contact details we ever see — the model extracted them and the
    visitor may have corrected them — so they are read from the last lead message, sent or not.
    """
    # JS: `[...messages].reverse().find(...)` copies first so the caller's array is untouched.
    # Python's list.reverse() mutates in place; reversed() is the non-destructive spelling.
    lead = next(
        (m for m in reversed(list(messages or [])) if m.role == "lead" and is_truthy(m.fields)),
        None,
    )
    if lead is None:
        return {}

    contact = {}
    for key in CONTACT_KEYS:
        value = lead.fields.get(key)
        if is_truthy(value) and str(value).strip():
            contact[key] = str(value).strip()
    return contact


def _to_wire(message: PanelMessage) -> WireMessage:
    """Lead cards carry fields rather than prose, so they are summarised rather than transcribed."""
    if message.role != "lead":
        return WireMessage(role=message.role, text=falsy_or(message.text, ""))
    fields = message.fields or {}
    interest = falsy_or(fields.get("interest"), "—")
    reason = falsy_or(fields.get("reason"), "—")
    return WireMessage(
        role="lead", text=f"[lead card shown] interest: {interest}; why now: {reason}"
    )


def build_digest(
    messages: Any,
    *,
    page: str = "",
    lead_sent: bool = False,
    share_authorized: bool = False,
    reason: str = "closed",
) -> OutboundDigest | None:
    """Returns None — JS `null` — for a visit that produced no visitor turn."""
    wire = tuple(w for w in (_to_wire(m) for m in (messages or [])) if w.text.strip())
    visitor_turns = sum(1 for m in wire if m.role == "user")
    if visitor_turns < MIN_VISITOR_TURNS:
        return None

    return OutboundDigest(
        messages=wire,
        contact=extract_contact(messages or []),
        page=page,
        lead_sent=lead_sent,
        share_authorized=share_authorized,
        reason=reason,
    )

File: src/components/chat/transport/chatDigest.js
/**
 * @file: src/components/chat/transport/chatDigest.js
 * @author: Stephen Boyett
 *
 * @description:
 *     Packs a Sol conversation for the team digest and posts it with sendBeacon, which is the
 *     only transport that survives the page being closed. Who receives it is decided server-side
 *     from `leadSent` / `shareAuthorized`. See functions/lib/CLAUDE.md § Conversation digests.
 *
 *     ⚠️ RETIRED 2026-08-26. Nothing imports this any more. Stephen replaced the
 *     per-conversation digest with one daily report (functions/lib/dailyReport.js), so the
 *     `sendChatDigest` endpoint was removed from index.js and the browser no longer beacons.
 *     Left on disk rather than deleted only because it is not in git; safe to delete once
 *     committed. See functions/lib/CLAUDE.md § The daily report replaced the digest.
 *
 * @See Also:
 *     functions/lib/digest.js
 *     src/components/chat/panel/ChatPanel.jsx
 *
 * ---
 * @Copyright © 2026 Cannasol Technologies LLC. All Rights Reserved.
 * ---
 */

export const DIGEST_URL = import.meta.env.DEV
  ? '/api/sendChatDigest'
  : 'https://us-central1-nano-kava-landing-page.cloudfunctions.net/sendChatDigest';

/** Matches the server's floor, so a visitor who opened Sol and said nothing sends nothing. */
export const MIN_VISITOR_TURNS = 1;

const CONTACT_KEYS = ['name', 'company', 'email', 'phone'];

/** How long a conversation sits idle before it counts as over. */
export const IDLE_TIMEOUT_MS = 3 * 60 * 1000;

/**
 * The lead card holds the best contact details we ever see — the model extracted them and the
 * visitor may have corrected them — so they are read from the last lead message, sent or not.
 */
export function extractContact(messages) {
  const lead = [...messages].reverse().find((m) => m.role === 'lead' && m.fields);
  if (!lead) return {};
  return CONTACT_KEYS.reduce((acc, key) => {
    const value = lead.fields[key];
    if (value && String(value).trim()) acc[key] = String(value).trim();
    return acc;
  }, {});
}

/** Lead cards carry fields rather than prose, so they are summarised rather than transcribed. */
const toWire = (message) => {
  if (message.role !== 'lead') return { role: message.role, text: message.text || '' };
  const fields = message.fields || {};
  return {
    role: 'lead',
    text: `[lead card shown] interest: ${fields.interest || '—'}; why now: ${fields.reason || '—'}`,
  };
};

export function buildDigest(messages, { page, leadSent = false, shareAuthorized = false, reason = 'closed' } = {}) {
  const wire = (messages || []).map(toWire).filter((m) => m.text.trim());
  const visitorTurns = wire.filter((m) => m.role === 'user').length;
  if (visitorTurns < MIN_VISITOR_TURNS) return null;

  return {
    messages: wire,
    contact: extractContact(messages || []),
    page: page || (typeof window !== 'undefined' ? window.location.pathname : ''),
    leadSent,
    shareAuthorized,
    reason,
  };
}

/**
 * Fire and forget. sendBeacon cannot report failure and we do not want it to: a digest is
 * telemetry for us, and nothing about the visitor's experience should depend on it.
 */
export function sendDigestBeacon(digest) {
  if (!digest || typeof navigator === 'undefined' || !navigator.sendBeacon) return false;
  try {
    // text/plain is CORS-safelisted; application/json preflights, and a preflight during unload is dropped.
    const blob = new Blob([JSON.stringify(digest)], { type: 'text/plain;charset=UTF-8' });
    return navigator.sendBeacon(DIGEST_URL, blob);
  } catch {
    return false;
  }
}
